apps/prairielearn/src/tests/groupExamRolePermissions.test.ts

```diff
@@ -0,0 +1,636 @@
+ */
+async function updateGroupRoles(
+  roleUpdates: any[],
+  groupRoles: any[],
+  studentUsers: StudentUser[],
+  csrfToken: string,
+  assessmentUrl: string,
+  $: cheerio.Root,
+): Promise<cheerio.Root> {
+  // Uncheck all of the inputs
+  const roleIds = groupRoles.map((role) => role.id);
+  const userIds = studentUsers.map((user) => user.user_id);
+  for (const roleId of roleIds) {
+    for (const userId of userIds) {
+      const elementId = `#user_role_${roleId}-${userId}`;
+      $('#role-select-form').find(elementId).removeAttr('checked');
+    }
+  }
+
+  let checkedBoxes = $('#role-select-form').find('tr').find('input:checked');
+  assert.lengthOf(checkedBoxes, 0, 'all checkboxes in role select form must be unchecked');
+
+  // Mark the checkboxes as checked
+  roleUpdates.forEach(({ roleId, groupUserId }) => {
+    $(`#user_role_${roleId}-${groupUserId}`).attr('checked', '');
+  });
+  checkedBoxes = $('#role-select-form').find('tr').find('input:checked');
+  assert.lengthOf(
+    checkedBoxes,
+    roleUpdates.length,
+    'all checkboxes in role select form must be checked',
+  );
+
+  // Grab IDs of checkboxes to construct update request
+  const checkedElementIds = {};
+  for (let i = 0; i < checkedBoxes.length; i++) {
+    checkedElementIds[checkedBoxes[i.toString()].attribs.id] = 'on';
+  }
+  const res = await fetch(assessmentUrl, {
+    method: 'POST',
+    body: new URLSearchParams({
+      __action: 'update_group_roles',
+      __csrf_token: csrfToken,
+      ...checkedElementIds,
+    }),
+  });
+  assert.isOk(res.ok, 'updating group roles should be successful');
+  return cheerio.load(await res.text());
+}
+
+async function getQuestionUrl(
+  courseInstanceUrl: string,
+  assessmentInstanceId: string,
+  questionId: string,
+): Promise<string> {
+  const result = await queryValidatedOneRow(
+    sql.select_instance_questions,
+    {
+      assessment_instance_id: assessmentInstanceId,
+      question_id: questionId,
+    },
+    QuestionIdSchema,
+  );
+  assert.isDefined(result.id);
+  return courseInstanceUrl + '/instance_question/' + result.id;
+}
+
+/**
+ * Validates and prepares a role-based group assessment with three users in a
+ * valid user configuration, then returns data for use in tests.
+ */
+async function prepareGroup() {
+  // Get exam assessment URL using ids from database
+  const assessmentResult = await queryOneRowAsync(sql.select_assessment, {
+    assessment_tid: GROUP_WORK_EXAM_TID,
+  });
+  assert.lengthOf(assessmentResult.rows, 1);
+  const assessmentId = assessmentResult.rows[0].id;
+  assert.isDefined(assessmentId);
+  const assessmentUrl = courseInstanceUrl + '/assessment/' + assessmentId;
+
+  // Generate three users
+  const studentUsers = await generateThreeStudentUsers();
+
+  // Get group roles
+  const groupRoles = await queryValidatedRows(
+    sql.select_assessment_group_roles,
+    {
+      assessment_id: assessmentId,
+    },
+    GroupRoleSchema,
```

Comment:
I'd add a complete schema for this to `lib/db-types.ts` rather than the one-off schema in this file.